src/controllers/midi/midisourceclock.h

```diff
@@ -0,0 +1,114 @@
+#ifndef MIDISOURCECLOCK_H
+#define MIDISOURCECLOCK_H
+
+#include <QList>
+
+#include "util/time.h"
+
+// Stub out the Time::elapsed function for easier testing.
+class MockableClock {
+  public:
+    virtual ~MockableClock() { };
+    virtual qint64 now() = 0;
+};
+
+class MixxxClock : public MockableClock {
+  public:
+    virtual qint64 now() {
+        return Time::elapsed();
+    }
+};
+
+// MidiSourceClock is not thread-safe, but is thread compatible using ControlObjects.
+// The MIDI thread will make calls into MidiSourceClock and then update two Control
+// Objects with the current reported BPM and last beat time.  The engine thread
+// can use those values to call a static function to calculate beat percentage
+// at any time in the future.  Time values are in nanoseconds for compatibility
+// with Time::elapsed().
+class MidiSourceClock {
+  public:
+    // The number of midi ticks per quarter note (1 beat in 4/4 time).
```

Comment:
again, pulseperquarter is the wording in the spec.  A quarter note in music isn't actually specific to 4/4 time signature!  a 3/4 measure still has "quarter notes."  Again, I want to stick to the common terminology rather than try to invent my own.  But I will unify tick->pulse